File: src/python/xerxes/runtime/history.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Lightweight chronological event log attached to every session.

:class:`HistoryEvent` is one immutable record; :class:`HistoryLog` is the
ordered append-only buffer the query engine writes to as turns, tool calls,
permission prompts, errors, and compaction events happen. The log feeds
``/history``, audit replays, and session persistence.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import Any


@dataclass(frozen=True)
class HistoryEvent:
    """One immutable history record.

    Attributes:
        kind: Short category tag (``"turn"``, ``"tool_call"``, ``"error"``,
            ``"permission_granted"``, ``"permission_denied"``, ...).
        title: Short headline describing the event.
        detail: Longer body text (often truncated to keep payloads small).
        timestamp: ISO-8601 timestamp captured at construction.
        metadata: Free-form extras stored alongside the record.
    """

    kind: str
    title: str
    detail: str = ""
    timestamp: str = field(default_factory=lambda: datetime.now().isoformat())
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class HistoryLog:
    """Append-only ordered list of :class:`HistoryEvent` records.

    Attributes:
        events: All events recorded so far, in insertion order.
    """

    events: list[HistoryEvent] = field(default_factory=list)

    def add(self, kind: str, title: str, detail: str = "", **metadata: Any) -> HistoryEvent:
        """Append a new :class:`HistoryEvent` and return it."""

        event = HistoryEvent(kind=kind, title=title, detail=detail, metadata=metadata)
        self.events.append(event)
        return event

    def add_tool_call(self, name: str, result_preview: str = "", duration_ms: float = 0.0) -> HistoryEvent:
        """Record a tool invocation; ``result_preview`` is truncated to 200 chars."""

        return self.add(
            kind="tool_call",
            title=name,
            detail=result_preview[:200],
            duration_ms=duration_ms,
        )

    def add_error(self, message: str, source: str = "") -> HistoryEvent:
        """Record an error event keyed by ``message`` with optional ``source``."""
        return self.add(kind="error", title=message, detail=source)

    def add_turn(self, model: str, in_tokens: int = 0, out_tokens: int = 0) -> HistoryEvent:
        """Record a completed turn with model and token counts."""

        return self.add(
            kind="turn",
            title=f"Turn completed ({model})",
            detail=f"in={in_tokens}, out={out_tokens}",
            model=model,
            in_tokens=in_tokens,
            out_tokens=out_tokens,
        )

    def add_permission(self, tool_name: str, granted: bool) -> HistoryEvent:
        """Record the result of a permission prompt for ``tool_name``."""

        status = "granted" if granted else "denied"
        return self.add(
            kind=f"permission_{status}",
            title=f"{tool_name}: {status}",
        )

    def filter_by_kind(self, kind: str) -> list[HistoryEvent]:
        """Return every event whose ``kind`` matches exactly."""
        return [e for e in self.events if e.kind == kind]

    def last(self, n: int = 10) -> list[HistoryEvent]:
        """Return the most recent ``n`` events in insertion order."""
        return self.events[-n:]

    @property
    def event_count(self) -> int:
        """Number of recorded events."""
        return len(self.events)

    def clear(self) -> None:
        """Drop every recorded event."""
        self.events.clear()

    def as_markdown(self) -> str:
        """Render the log as a Markdown bullet list, one line per event."""

        lines = ["# Session History", "", f"Events: {self.event_count}", ""]
        for event in self.events:
            ts = event.timestamp[:19]
            detail = f" — {event.detail}" if event.detail else ""
            lines.append(f"- [{ts}] **{event.kind}**: {event.title}{detail}")
        return "\n".join(lines)

    def as_dicts(self) -> list[dict[str, Any]]:
        """Serialise every event as a JSON-friendly dict for persistence."""

        return [
            {
                "kind": e.kind,
                "title": e.title,
                "detail": e.detail,
                "timestamp": e.timestamp,
                **e.metadata,
            }
            for e in self.events
        ]


__all__ = [
    "HistoryEvent",
    "HistoryLog",
]

File: xerxes/src/runtime/history.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

export interface HistoryEventOptions {
  readonly detail?: string
  readonly kind: string
  readonly metadata?: Readonly<Record<string, unknown>>
  readonly timestamp?: string
  readonly title: string
}

/** Immutable chronological runtime event. */
export class HistoryEvent {
  readonly detail: string
  readonly kind: string
  readonly metadata: Readonly<Record<string, unknown>>
  readonly timestamp: string
  readonly title: string

  constructor(options: HistoryEventOptions) {
    this.kind = textValue(options.kind, 'kind')
    this.title = textValue(options.title, 'title')
    this.detail = textValue(options.detail ?? '', 'detail')
    this.timestamp = timestampValue(options.timestamp ?? new Date().toISOString())
    this.metadata = freezeRecord(options.metadata ?? {})
    Object.freeze(this)
  }

  toRecord(): Readonly<Record<string, unknown>> {
    return Object.freeze({
      kind: this.kind,
      title: this.title,
      detail: this.detail,
      timestamp: this.timestamp,
      ...cloneRecord(this.metadata),
    })
  }
}

export interface HistoryLogOptions {
  readonly now?: () => Date
}

/** Ordered append-only session history suitable for display and persistence. */
export class HistoryLog {
  private readonly clock: () => Date
  private readonly ledger: HistoryEvent[] = []

  constructor(options: HistoryLogOptions = {}) {
    this.clock = options.now ?? (() => new Date())
  }

  get events(): readonly HistoryEvent[] {
    return this.ledger.map(event => copyEvent(event))
  }

  get eventCount(): number {
    return this.ledger.length
  }

  add(
    kind: string,
    title: string,
    detail = '',
    metadata: Readonly<Record<string, unknown>> = {},
  ): HistoryEvent {
    const event = new HistoryEvent({
      kind,
      title,
      detail,
      metadata,
      timestamp: this.nowTimestamp(),
    })
    this.ledger.push(event)
    return copyEvent(event)
  }

  addToolCall(name: string, resultPreview = '', durationMs = 0): HistoryEvent {
    return this.add('tool_call', name, resultPreview.slice(0, 200), { duration_ms: durationMs })
  }

  addError(message: string, source = ''): HistoryEvent {
    return this.add('error', message, source)
  }

  addTurn(model: string, inputTokens = 0, outputTokens = 0): HistoryEvent {
    return this.add(
      'turn',
      'Turn completed (' + model + ')',
      'in=' + inputTokens + ', out=' + outputTokens,
      { model, in_tokens: inputTokens, out_tokens: outputTokens },
    )
  }

  addPermission(toolName: string, granted: boolean): HistoryEvent {
    const status = granted ? 'granted' : 'denied'
    return this.add('permission_' + status, toolName + ': ' + status)
  }

  filterByKind(kind: string): HistoryEvent[] {
    return this.ledger.filter(event => event.kind === kind).map(copyEvent)
  }

  last(count = 10): HistoryEvent[] {
    if (!Number.isInteger(count)) throw new RangeError('count must be an integer')
    return this.ledger.slice(-Math.max(0, count)).map(copyEvent)
  }

  clear(): void {
    this.ledger.length = 0
  }

  asMarkdown(): string {
    const lines = ['# Session History', '', 'Events: ' + this.eventCount, '']
    for (const event of this.ledger) {
      const detail = event.detail ? ' — ' + event.detail : ''
      lines.push('- [' + event.timestamp.slice(0, 19) + '] **' + event.kind + '**: ' + event.title + detail)
    }
    return lines.join('\n')
  }

  asDicts(): ReadonlyArray<Readonly<Record<string, unknown>>> {
    return this.ledger.map(event => event.toRecord())
  }

  private nowTimestamp(): string {
    const now = this.clock()
    if (!(now instanceof Date) || Number.isNaN(now.valueOf())) throw new RangeError('now must return a valid Date')
    return now.toISOString()
  }
}

function copyEvent(event: HistoryEvent): HistoryEvent {
  return new HistoryEvent({
    kind: event.kind,
    title: event.title,
    detail: event.detail,
    timestamp: event.timestamp,
    metadata: event.metadata,
  })
}

function textValue(value: unknown, name: string): string {
  if (typeof value !== 'string') throw new TypeError(name + ' must be a string')
  return value
}

function timestampValue(value: unknown): string {
  if (typeof value !== 'string' || Number.isNaN(new Date(value).valueOf())) {
    throw new RangeError('timestamp must be a valid ISO timestamp')
  }
  return value
}

function cloneRecord(record: Readonly<Record<string, unknown>>): Readonly<Record<string, unknown>> {
  return freezeRecord(record)
}

function freezeRecord(record: Readonly<Record<string, unknown>>): Readonly<Record<string, unknown>> {
  const copy: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(record)) copy[key] = freezeValue(value)
  return Object.freeze(copy)
}

function freezeValue(value: unknown): unknown {
  if (Array.isArray(value)) return Object.freeze(value.map(freezeValue))
  if (value instanceof Date) return value.toISOString()
  if (value && typeof value === 'object') return freezeRecord(value as Record<string, unknown>)
  return value
}
